Add "run" command

JordanMartinez opened the pull request:
Resolves #131 

PR checklist:
- [x] Added a test for the contribution (if applicable)
    - Not sure whether this is sufficient. I'm not sure how to write a test in this situation.
- [ ] Updated the version number [in `package.yaml`](https://github.com/spacchetti/spago/blob/master/package.yaml#L2) according to SemVer (`0.x.y.z`, `x` changes on breaking changes, `y` on additions, `z` on patches)
    - Haven't done this, but will if desired. (I"ve noticed that others haven't done this as well, so is this PR template old/outdated?)
- [x] Added some example of the new feature to the `README`

f-f commented on app/Spago/Build.hs:
```diff
@@ -69,6 +69,18 @@ test maybeModuleName maybeLimit paths passthroughArgs = do
+    T.ExitFailure n -> die $ "Run failed: " <> T.repr n
+  where
+    moduleName = fromMaybe (Purs.ModuleName "Main") maybeModuleName
+    cmd = "node -e \"require('./output/" <> Purs.unModuleName moduleName <> "').main()\""
```
Since these commands are doing basically the same thing, we can extract this common code. I.e. something like:

```hs
run :: Maybe Purs.ModuleName -> Maybe Int -> [Purs.SourcePath] -> [Purs.ExtraArg] -> IO ()
run = runWithNode (Purs.ModuleName "Main")

test :: Maybe Purs.ModuleName -> Maybe Int -> [Purs.SourcePath] -> [Purs.ExtraArg] -> IO ()
test = runWithNode (Purs.ModuleName "Test.Main")

-- | Run the project with node: compile and run with the provided ModuleName
--   (or the default one if that's missing)
runWithNode
  :: Purs.ModuleName
  -> Maybe Purs.ModuleName
  -> Maybe Int
  -> [Purs.SourcePath]
  -> [Purs.ExtraArg]
  -> IO ()
runWithNode defaultModuleName maybeModuleName maybeLimit paths passthroughArgs = do
  build maybeLimit paths passthroughArgs
  T.shell cmd T.empty >>= \case
    T.ExitSuccess   -> pure ()
    T.ExitFailure n -> die $ "Running failed, exit code: " <> T.repr n
  where
    moduleName = fromMaybe defaultModuleName maybeModuleName
    cmd = "node -e \"require('./output/" <> Purs.unModuleName moduleName <> "').main()\""

```

JordanMartinez replied:
I'll update it to do that.

JordanMartinez replied:
I realized I needed to add a `Maybe String` and `String argument for the success/failure messages to keep the messages the same throughout both commands. However, I don't know whether those arguments' types should be `String`, which I know is an alias for `[Char]`, or `Text`.

f-f replied:
As you might have figured from the CI run, all the strings should be `Text`

You can compile the program locally with `stack build --fast` so you don't have to wait for CI

Also you can get a repl with `stack repl` so you can ask for the types of things, e.g. `:t echo` returns `echo :: Text -> IO ()`)

f-f commented:
@JordanMartinez re the version number: yeah I put that checklist item in the PR template, but never actually followed it myself 🙁 So I'd say it can be removed!

JordanMartinez commented:
> re the version number: yeah I put that checklist item in the PR template, but never actually followed it myself slightly_frowning_face So I'd say it can be removed!

Should I remove that in this PR?

Ah, I forgot that Haskell's "unit" is "()" (why the build failed)

f-f commented:
@JordanMartinez 

> Should I remove that in this PR?

This PR or another PR, both are good :)

> Not sure whether this is sufficient. I'm not sure how to write a test in this situation.

This is good, but a better test would be to test this in the same way we [test the `test` command](https://github.com/spacchetti/spago/blob/ef0bd85fe0239aba8ea418d07ee2bffc237dabd7/test/spago-test.py#L129), with a [fixture](https://github.com/spacchetti/spago/blob/ef0bd85fe0239aba8ea418d07ee2bffc237dabd7/test/fixtures/test-output.txt)

JordanMartinez commented:
> This is good, but a better test would be to test this in the same way we test the test command, with a fixture

Saw that. I wasn't sure what the output would be. 

For example, `test`'s output is:
```
🍝
You should add some tests.
Installing 3 dependencies.
"prelude" already installed
"console" already installed
"effect" already installed
Installation complete.
Build succeeded.
Tests succeeded.

```
For `run`, the 🍝 should be the last line outputted. Since they use the same command, it seems like it'll end up looking like this:
```
🍝
Installing 3 dependencies.
"prelude" already installed
"console" already installed
"effect" already installed
Installation complete.
Build succeeded.

```
when it should look like this:
```
Installing 3 dependencies.
"prelude" already installed
"console" already installed
"effect" already installed
Installation complete.
Build succeeded.
🍝
```
Right?

f-f commented:
@JordanMartinez indeed. Also it looks like the tests have been passing lately, even though that's not supposed to be the output of the `test` command (the lines "$dependency already installed" are not there anymore), so probably the fixtures are not actually being checked.. 🤔

The test you added is good then, I'll have to look better into the fixture checking!

f-f commented on app/Spago/Build.hs:
```diff
@@ -57,16 +57,35 @@ repl sourcePaths passthroughArgs = do
 --   (or the provided module name) with node
 test :: Maybe Purs.ModuleName -> Maybe Int -> [Purs.SourcePath] -> [Purs.ExtraArg] -> IO ()
-test maybeModuleName maybeLimit paths passthroughArgs = do
+test maybeModuleName successMessage failureMessage maybeLimit paths passthroughArgs =
```
```suggestion
test =
```

JordanMartinez replied:
Ha! Silly me...

f-f commented on app/Spago/Build.hs:
```diff
@@ -57,16 +57,35 @@ repl sourcePaths passthroughArgs = do
+-- | Run the project: compile and run "Main"
+--   (or the provided module name) with node
+run :: Maybe Purs.ModuleName -> Maybe Int -> [Purs.SourcePath] -> [Purs.ExtraArg] -> IO ()
+run maybeModuleName successMessage failureMessage maybeLimit paths passthroughArgs =
```
```suggestion
run =
```

f-f commented on app/Spago/Build.hs:
```diff
@@ -57,16 +57,35 @@ repl sourcePaths passthroughArgs = do
+-- | Run the project with node: compile and run with the provided ModuleName
+--   (or the default one if that's missing)
+runWithNode :: Purs.ModuleName
+            -> Maybe String
```
I'd be fine just with having `pure ()` for success and `Run failed` in case of failure so we could avoid passing this many arguments here, but if we want to then they should be `Text`

JordanMartinez replied:
I'd rather keep the `Text`. Reading the last line of a `test` command as "Tests succeeded." seems like a better UI than just a blank message.

JordanMartinez commented:
> Also it looks like the tests have been passing lately, even though that's not supposed to be the output of the test command (the lines "$dependency already installed" are not there anymore), so probably the fixtures are not actually being checked.. thinking

They aren't by the looks of it. I deleted the content of the text fixture and ran the tests. It said the `test` command succeeded and failed on the `run` command (because I haven't rebuilt spago)

JordanMartinez commented:
Once the fixture stuff gets figured out, I could rebase these changes on top of it and add a `run-output.txt` fixture that is similar to the `test-output.txt` fixture.

JordanMartinez commented:
Sigh...

Travis CI:
```
/home/ubuntu/spago/app/Spago/Build.hs:83:24: error:
    • Couldn't match type ‘()’ with ‘IO ()’
      Expected type: Maybe T.Text -> IO ()
        Actual type: Maybe T.Text -> ()
    • The function ‘fromMaybe’ is applied to three arguments,
      its type is ‘a0 -> Maybe a0 -> f0 ()’,
      it is specialized to ‘(Maybe T.Text -> ())
                            -> Maybe (Maybe T.Text -> ()) -> Maybe T.Text -> ()’
      In the expression: fromMaybe (pure ()) echo maybeSuccessMessage
      In a case alternative:
          T.ExitSuccess -> fromMaybe (pure ()) echo maybeSuccessMessage
   |
83 |     T.ExitSuccess   -> fromMaybe (pure ()) echo maybeSuccessMessage
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/home/ubuntu/spago/app/Spago/Build.hs:83:44: error:
    • Couldn't match expected type ‘Maybe (Maybe T.Text -> ())’
                  with actual type ‘T.Text -> IO ()’
    • Probable cause: ‘echo’ is applied to too few arguments
      In the second argument of ‘fromMaybe’, namely ‘echo’
      In the expression: fromMaybe (pure ()) echo maybeSuccessMessage
      In a case alternative:
          T.ExitSuccess -> fromMaybe (pure ()) echo maybeSuccessMessage
   |
83 |     T.ExitSuccess   -> fromMaybe (pure ()) echo maybeSuccessMessage
   |                                            ^^^^
Completed 32 action(s).
```
I'm waiting for `stack build --fast` to finish and I'll troubleshoot from there.

f-f commented:
@JordanMartinez sorry, I missed that when reviewing; you probably wanted to write
```hs
fromMaybe (pure ()) (echo <$> maybeSuccessMessage)
```

JordanMartinez commented:
Ah.... duh. Yeah, that's it.

JordanMartinez commented:
Great. This build now. Let me know when that test fixture gets fixed.

Also, I wonder whether the `run` command's docs should also mention that arguments can be passed through to the underlying program. Perhaps I can fix that when rebasing, so that this doesn't trigger yet another 30-min CI build?

f-f commented on README.md:
```diff
@@ -428,6 +428,18 @@ $ node -e "console.log(require('./index).main)"
+$ spago run --main ModulePath.To.Main
+
+# And pass arguments through to your program
+$ spago run --main ModulePath.To.Main -- --some-option "argument"
```
These additional arguments are passed to `purs compile`

JordanMartinez replied:
Hmm... Well, then... This command does not work "akin to `pulp run` then, AFAIK.

Is that possibility (pass args to `purs compile` and other args to the actual program) desired?

f-f replied:
If we want to make it work like `pulp run` then it's desired. Though I think we could take care of that in another PR?

f-f commented:
@JordanMartinez the fix for #138 went in, I updated this PR with the latest changes from master

JordanMartinez commented:
I might get to this later tonight or perhaps two days from now (busy tomorrow). Still, that's good to know that the fixture checking is working now.

f-f commented:
@JordanMartinez no worries!
(you'll note that now that the `--watch` flag went in you might want to use it in the `run` command too)

JordanMartinez commented:
I had some time to day, so I've 
- merged this branch into the current master
- updated the ReadMe example to show that CLI args are passed to `purs compile`, not the actual program. (We can take care of that issue in another PR as I think it will require some refactoring.)
- updated the run command's test fixture to be what I think it should be. However, I'm not sure whether it will build correctly.

f-f commented:
@JordanMartinez I did some tweaks to the install procedure in #125, and now there's an `Installing 3 dependencies.` too much in the fixture for `run`

Other than that everything looks great!

f-f commented on README.md:
```diff
@@ -214,11 +214,11 @@ $ spago install my-new-package another-package
 
-We have a `overrides` record in `packages.dhall` just for that!  
-And in this case we override the `repo` key with the local path of the package.  
+We have a `overrides` record in `packages.dhall` just for that!
```
```suggestion
We have a `overrides` record in `packages.dhall` just for that!  
```

f-f commented on README.md:
```diff
@@ -214,11 +214,11 @@ $ spago install my-new-package another-package
-We have a `overrides` record in `packages.dhall` just for that!  
-And in this case we override the `repo` key with the local path of the package.  
+We have a `overrides` record in `packages.dhall` just for that!
+And in this case we override the `repo` key with the local path of the package.
```
```suggestion
And in this case we override the `repo` key with the local path of the package.  
```

f-f commented on README.md:
````diff
@@ -255,7 +255,7 @@ Installing "enums"
 ```
 
-Let's now say that we test that our fix works, and we are ready to Pull Request the fix.  
+Let's now say that we test that our fix works, and we are ready to Pull Request the fix.
````
```suggestion
Let's now say that we test that our fix works, and we are ready to Pull Request the fix.  
```

f-f commented on test/fixtures/run-output.txt:
```diff
@@ -0,0 +1,4 @@
+🍝
+Installing 3 dependencies.
```
```suggestion
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
-- **`psc-package-local-setup`**: this command creates a `packages.dhall` file in your project, 
-  that points to the most recent package-set, and lets you override and add  arbitrary packages.  
+- **`psc-package-local-setup`**: this command creates a `packages.dhall` file in your project,
+  that points to the most recent package-set, and lets you override and add  arbitrary packages.
```
```suggestion
  that points to the most recent package-set, and lets you override and add  arbitrary packages.  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 
 So this is the reason why if you or one of your dependencies need to depend on some "native"
-packages, you should run the appropriate package-manager for that (e.g. npm).  
+packages, you should run the appropriate package-manager for that (e.g. npm).
```
```suggestion
packages, you should run the appropriate package-manager for that (e.g. npm).  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 
 Another big problem is that the JS backend is not the only backend around. For example,
-PureScript has a [C backend][purec] and an [Erlang backend][purerl] among the others.  
+PureScript has a [C backend][purec] and an [Erlang backend][purerl] among the others.
```
```suggestion
PureScript has a [C backend][purec] and an [Erlang backend][purerl] among the others.  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 While these scenarios are common, they are also really hard to support.
 In fact, it might be that a certain NPM package in your transitive dependencies
-would only support the browser, or only node. Should `spago` warn about that?  
+would only support the browser, or only node. Should `spago` warn about that?
```
```suggestion
would only support the browser, or only node. Should `spago` warn about that?  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 The gist of it is that the PureScript apps in the repo are built with `spago build`
 (look in the `package.json` for it), but all the React components can be imported from
-JS apps as well, given that proper modules are built out of the PS sources.  
+JS apps as well, given that proper modules are built out of the PS sources.
```
```suggestion
JS apps as well, given that proper modules are built out of the PS sources.  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
-For an example of this in a "production setting" you can take a look at [affresco][affresco].  
-It is a PureScript monorepo of React-based components and apps.  
+For an example of this in a "production setting" you can take a look at [affresco][affresco].
+It is a PureScript monorepo of React-based components and apps.
```
```suggestion
It is a PureScript monorepo of React-based components and apps.  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 
-For an example of this in a "production setting" you can take a look at [affresco][affresco].  
-It is a PureScript monorepo of React-based components and apps.  
+For an example of this in a "production setting" you can take a look at [affresco][affresco].
```
```suggestion
For an example of this in a "production setting" you can take a look at [affresco][affresco].  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 starting point is the [TodoMVC app with `react-basic`][todomvc].
 You can see in its `package.json` that a "production build" is just
-`spago build && parcel build index.html`.  
+`spago build && parcel build index.html`.
```
```suggestion
`spago build && parcel build index.html`.  
```

f-f commented on README.md:
```diff
@@ -532,64 +543,64 @@ the `dependencies` list in your `spago.dhall`.
 No. We only take care of PureScript land. In particular, `make-module` will do the
 most we can do on the PureScript side of things (dead code elimination), but will
-leave the `require`s still in.  
+leave the `require`s still in.
```
```suggestion
leave the `require`s still in.  
```

f-f commented on README.md:
```diff
@@ -512,8 +523,8 @@ Yees, however:
   packages versions when needed, leading everyone to make their own package-set, which
-  then goes unmaintained, etc.  
-  Of course you can use the package-set-local-setup to solve this issue, but this is 
+  then goes unmaintained, etc.
```
```suggestion
  then goes unmaintained, etc.  
```

f-f commented on README.md:
````diff
@@ -388,7 +388,7 @@ We can build the project and its dependencies by running:
 ```
 
-This is just a thin layer above the PureScript compiler command `purs compile`.  
+This is just a thin layer above the PureScript compiler command `purs compile`.
````
```suggestion
This is just a thin layer above the PureScript compiler command `purs compile`.  
```

f-f commented on README.md:
```diff
@@ -376,7 +376,7 @@ $ spago freeze
 
 However, this is a pretty rare situation and in principle it should not happen, and when
-it happens it might not be secure to run the above command.  
+it happens it might not be secure to run the above command.
```
```suggestion
it happens it might not be secure to run the above command.  
```

f-f commented on README.md:
```diff
@@ -274,11 +274,11 @@ let overrides =
 
 If a package is not in the upstream package-set, you can add it in a similar way,
-by changing the `additions` record in the `packages.dhall` file.  
+by changing the `additions` record in the `packages.dhall` file.
```
```suggestion
by changing the `additions` record in the `packages.dhall` file.  
```

f-f commented on README.md:
```diff
@@ -274,11 +274,11 @@ let overrides =
 **Note**: currently only "branches" and "tags" work as a `version`, and tags are
 recommended over branches (as for example if you push new commits to a branch,
-`spago` won't pick them up unless you delete the `.spago` folder).  
+`spago` won't pick them up unless you delete the `.spago` folder).
```
```suggestion
`spago` won't pick them up unless you delete the `.spago` folder).  
```

f-f commented on README.md:
```diff
@@ -214,11 +214,11 @@ $ spago install my-new-package another-package
 
 Let's say I'm a user of the `simple-json` package. Now, let's say I stumble upon a bug
-in there, but thankfully I figure how to fix it. So I clone it locally and add my fix.  
+in there, but thankfully I figure how to fix it. So I clone it locally and add my fix.
```
```suggestion
in there, but thankfully I figure how to fix it. So I clone it locally and add my fix.  
```

f-f commented on README.md:
```diff
@@ -92,13 +92,13 @@ This last command will create a bunch of files:
 Let's take a look at the two [Dhall][dhall] configuration files that `spago` requires:
 - `packages.dhall`: this file is meant to contain the *totality* of the packages
-  available to your project (that is, any package you might want to import).  
+  available to your project (that is, any package you might want to import).
```
```suggestion
  available to your project (that is, any package you might want to import).  
```

f-f commented on README.md:
```diff
@@ -92,13 +92,13 @@ This last command will create a bunch of files:
 
 Convention note: `spago` expects your source files to be in `src/` and your
-test files in `test/`.  
+test files in `test/`.
```
```suggestion
test files in `test/`.  
```

f-f commented on README.md:
```diff
@@ -49,7 +49,7 @@ PureScript package manager and build tool powered by [Dhall][dhall] and
 ## What does all of this mean?
 
-`spago` aims to tie together the UX of developing a PureScript project.  
+`spago` aims to tie together the UX of developing a PureScript project.
```
```suggestion
`spago` aims to tie together the UX of developing a PureScript project.  
```

JordanMartinez commented:
> It looks like your editor removed some markdown syntax (the double space adds a newline without making a new paragraph), so added suggestions in all the relevant places slightly_smiling_face

Hmm... I didn't know that syntax did that. I always added the newline.

And yeah, my editor did this the first time I submitted a PR, but I caught it beforehand and fixed its mistakes. I missed it this time.

This PR is ready to be merged:
- fixed the "  " removals in a previous commit
- updated the run fixture to remove the "Installing 3 dependencies." line so that it passes
- merged into master

JordanMartinez commented:
Note: this should probably be merged before #145 since that will affect what should be included in the run fixture file.

f-f commented on README.md:
```diff
@@ -426,10 +435,10 @@ You should add some tests.
 As with the `build` and `test` commands, you can add custom source paths
 to load, and pass options to the underlying `purs repl` by just putting
-them after `--`.  
+them after `--`.
```
```suggestion
them after `--`.  
```

f-f commented on README.md:
````diff
@@ -426,10 +435,10 @@ You should add some tests.
 ```
 
-And last but not least, you can spawn a PureScript repl!  
+And last but not least, you can spawn a PureScript repl!
````
```suggestion
And last but not least, you can spawn a PureScript repl!  
```

f-f commented on README.md:
```diff
@@ -358,7 +358,7 @@ Done. Updating the local package-set file..
 
 If you wish to detach from tags for your package-set, you can of course point it to a
-specific commit.  
+specific commit.
```
```suggestion
specific commit.  
```

f-f commented:
@JordanMartinez we can merge this before #145, but it looks like the whitespaced are still being removed?

JordanMartinez commented:
I started fixing that and came across quite a surprise: as soon as I fix a second line and save the file, the editor (Atom) will remove the two spaces from the first line I fixed. :-/
I think I'll just do this via GH

f-f commented:
@JordanMartinez could you also approve the previous suggested changes?

JordanMartinez commented:
@f-f Sorry for becoming a bottleneck. I believe I've addressed everything you've mentioned thus far.

f-f commented:
@JordanMartinez no worries! And thanks a lot :slightly_smiling_face:
